File: internal/core/ports/events.go
// define event payloads and emitter contracts.
// internal/core/ports/events.go
package ports

import "github.com/MadeByDoug/wls-chatbot/internal/features/chat/domain"

// ChatEvent is the wrapper for all events sent to the frontend.
type ChatEvent struct {
	Type           string      `json:"type"`
	ConversationID string      `json:"conversationId"`
	MessageID      string      `json:"messageId,omitempty"`
	Timestamp      int64       `json:"ts"`
	Payload        interface{} `json:"payload,omitempty"`
}

// StreamChunkPayload is the payload for stream chunk events.
type StreamChunkPayload struct {
	BlockIndex int                   `json:"blockIndex"`
	Content    string                `json:"content"`
	IsDone     bool                  `json:"isDone"`
	Metadata   *chat.MessageMetadata `json:"metadata,omitempty"`
	Error      string                `json:"error,omitempty"`
	StatusCode int                   `json:"statusCode,omitempty"`
}

// ConversationTitlePayload carries a conversation title update.
type ConversationTitlePayload struct {
	Title string `json:"title"`
}

// ToastPayload describes a toast notification.
type ToastPayload struct {
	Type    string `json:"type,omitempty"`
	Title   string `json:"title,omitempty"`
	Message string `json:"message,omitempty"`
}

// ChatEmitter publishes chat-related events to the frontend.
type ChatEmitter interface {
	EmitChatEvent(event ChatEvent)
}

// ProviderEmitter publishes provider-related events to the frontend.
type ProviderEmitter interface {
	EmitProvidersUpdated()
}

// CatalogEmitter publishes catalog-related events to the frontend.
type CatalogEmitter interface {
	EmitCatalogUpdated()
}

// ToastEmitter publishes toast notifications to the frontend.
type ToastEmitter interface {
	EmitToast(payload ToastPayload)
}

// Emitter bundles all event emitters used by policies.
type Emitter interface {
	ChatEmitter
	ProviderEmitter
	CatalogEmitter
	ToastEmitter
}

File: frontend/src/types/events.ts
/**
 * define backend-to-frontend event payload shapes.
 * frontend/src/types/events.ts
 */

import type { Message, MessageMetadata } from './wails';

/**
 * enumerate supported chat event types.
 */
export type ChatEventType =
    | 'chat.message'
    | 'chat.stream.start'
    | 'chat.stream.chunk'
    | 'chat.stream.error'
    | 'chat.stream.complete'
    | 'chat.conversation.title';

/**
 * describe the payload for streaming content chunks.
 */
export interface StreamChunkPayload {
    blockIndex: number;
    content: string;
    isDone: boolean;
    metadata?: MessageMetadata;
    error?: string;
    statusCode?: number;
}

/**
 * describe a conversation title update payload.
 */
export interface ConversationTitlePayload {
    title: string;
}

/**
 * wrap chat events with metadata and typed payloads.
 */
export interface ChatEvent<TPayload = Message | StreamChunkPayload | ConversationTitlePayload | Record<string, unknown>> {
    type: ChatEventType;
    conversationId: string;
    messageId?: string;
    ts: number;
    payload?: TPayload;
}
